### PR TITLE
docs add entity-service to contributor docs

### PR DESCRIPTION
**Pull request by Boegie19**
### What does it do?

Add 2 checklists for working with decorators

### Why is it needed?

To ensure that if new decorators are all following the decorator rules so all decorators down stream and up stream will work properly

### CODE REVIEW

**Comment by joshuaellis** [comment]
Before i "correct" the labels, can i check the intent of your PR? To add some documentation to our contributor docs? :)

File: docs/docs/core/entity-service.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+---
+title: Entity Service
+slug: /entity-service
```
Comment (joshuaellis) [inline docs/docs/core/entity-service.md:3]: IMO this should be `core/database/entity-service` because afaik it relates to the `database` package. wdyt? 

cc @Marc-Roig

Comment (Marc-Roig) [reply docs/docs/core/entity-service.md:3]: helloo, 
the entity service lives in core actually , not in the database package
`packages/core/strapi/lib/services/entity-service/index.js`

File: docs/docs/core/entity-service.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+tags:
+  - Strapi Service
+  - Entity Service
+---
+
+# Entity Service
+The Entity Service is a wrapper around strapi.db.query
```
Comment (Marc-Roig) [inline docs/docs/core/entity-service.md:10]: I would say something similar as in our docs, 

Strapi provides an Entity Service API, built on top of the Query Engine API (Database Layer). The Entity Service is the layer that handles Strapi's complex data structures like components and dynamic zones, and uses the Query Engine API under the hood to execute database queries.

I think it is important to mention that everything core bussiness related to strapi is inside the entity service, like compos, dz and webhooks.

File: docs/docs/core/entity-service.md (hunk 1 above)
Comment (joshuaellis) [reply docs/docs/core/entity-service.md:3]: ah okay, so maybe it should be `core/strapi/entity-service`? because it still lives in a package?

File: docs/docs/core/entity-service.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+This wrapper uses the decorator pattern so plugins and strapi core packages can insert decorators to change the output of the entity service function
+Examples are: plugins/i18n and the core/upload
+
+## Checklist writing decorator
```
Comment (Marc-Roig) [inline docs/docs/core/entity-service.md:14]: I would call this recommendations, and I would give some code examples. 
The developers using the decorator might want to ignore this even if they are good practices.

File: docs/docs/core/entity-service.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+- You are always required to call service.(function).call() in your decorator function
+- Always return the `opts`/`options` you get from the parent decorator to service into `service.(function).call()`. you are allowed to add your on stuff to it but the parent opts should always be respected.
+- when using `service.(function).call()` ensure the first parameter is always `this` to not break the call chain
+- Your function should always return the output of `service.(function).call()` how ever you are allowed to add stuff to it.
+- when using any wrap functions do not call them in your decorator they are called by the Main Decorator only
```
Comment (Marc-Roig) [inline docs/docs/core/entity-service.md:19]: As previously said, I would phrase this as recommendations:
- Call `service.METHOD.call()` in your decorator, so upstream and downstream decorators are also executed.
```js
const decorator = (service) => ({
    async findMany(uid, opts = {}) {
        const result =  service.findMany.call(this, uid, opts);
        // ... update result
		return result;
	}
}
```
- Pass all arguments of the decorated method into `service.METHOD.call()`. Change, add and mutate as necessary,  but take into consideration other decorators might need those arguments too.
- Use `service.(function).call()` to call the downstream decorators.


I would prefer to have a table with good and bad examples too.

<table>
<tr>
 <td>GOOD 🔥
 <td>BAD 👎
<tr>
 <td>
```
 some code
 ```
 <td>
```
 some code
 ```
</table>

File: docs/docs/core/entity-service.md (hunk 1 above)
Comment (Marc-Roig) [reply docs/docs/core/entity-service.md:3]: indeed! it could be in `core/strapi/entity-service`

Comment (Boegie19) [reply docs/docs/core/entity-service.md:3]: `core/strapi/entity-service` it is

File: docs/docs/core/entity-service.md (hunk 3 above)
Comment (Boegie19) [reply docs/docs/core/entity-service.md:14]: Marc ignoring those means literally breaking other decorators above or below it

Comment (Marc-Roig) [reply docs/docs/core/entity-service.md:14]: and that is for the plugin developer to decide. I would explain why it is a recommendation, and what you might be breaking. But I don't think it's necessary to go that far and say it's "mandatory".

File: docs/docs/core/entity-service.md (hunk 1 above)
Comment (Boegie19) [reply docs/docs/core/entity-service.md:3]: @joshuaellis Then we should probably do the same for the event-hub

File: docs/docs/core/entity-service.md (hunk 3 above)
Comment (Boegie19) [reply docs/docs/core/entity-service.md:14]: I respect your wishes.

But I would still say mandatory since this are contributor docs for strapi aka for strapi maintained plugins so strapi desiding to break other plugins from themselfs would be a bad idea.

Aka what you are basicly telling me is to say sure you can break our I18n plugin with review workflows or your new url signing

But I will make the change to make it recomandations since you asked nicely and it are strapi's docs not mine
@Marc-Roig

File: docs/docs/core/entity-service.md (hunk 4 above)
Comment (Boegie19) [reply docs/docs/core/entity-service.md]: @Marc-Roig What would be best having a good and bad example for every point directly bellow it?

File: docs/docs/core/entity-service.md (hunk 1 above)
Comment (joshuaellis) [reply docs/docs/core/entity-service.md]: If you could move it around, i'd be really grateful :)

Comment (Boegie19) [reply docs/docs/core/entity-service.md]: I moved the location of the file to core/strapi/entity-service
Changing the slug breaks breaks the sidebar and I can't fix that since the location would be 
`core/strapi/entity-service/core/strapi/entity-service` what does not work

File: docs/docs/core/strapi/entity-service.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
+---
+title: Entity Service
```
Comment (joshuaellis) [inline docs/docs/core/strapi/entity-service.md:2]:
```suggestion
title: Entity Service
slug: /strapi/entity-service
```

This should give us the desired URL :)

File: docs/docs/core/strapi/entity-service.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+tags:
+  - Strapi Service
+  - Entity Service
```
Comment (joshuaellis) [inline docs/docs/core/strapi/entity-service.md:5]:
```suggestion
tags:
  - strapi
  - service
  - entity-service
```

This just brings them inline with how we've done tags elsewhere in the docs 👍🏼

File: docs/docs/core/strapi/entity-service.md (hunk 5 above)
Comment (Boegie19) [reply docs/docs/core/strapi/entity-service.md:2]: @joshuaellis currently put it in a strapi folder and also added a strapi section to the docs

Comment (joshuaellis) [reply docs/docs/core/strapi/entity-service.md:2]: no, fine is both, this is how we're doing it everywhere else iirc

**Review by Umer27 — COMMENTED** [review]
entity-service added

File: docs/docs/core/strapi/entity-service.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+
+SOME CODE  EXAMPLE
+
+
+### decorators
+When writing a decorator, please make sure you follow this checklist:
+- You should call `service.(function).call()` in your decorator function. the consequence of not doing this is that you are not calling the decorators bellow this decorator. What would mean you ignore them.
+- You are recommend to add your changes to the output of `service.(function).call()` so you can return it with the output of the lower laying decorators.
+- It is best to use `opts`/`options` you get from the parent decorator into your service call into `service.(function).call()`. You can add stuff to it but please remember to not change the options you get. Since change them would have as an effect that it could break the decorator above it.
```
Comment (markkaylor) [inline docs/docs/core/strapi/entity-service.md:24]:
```suggestion
- It is best to use the `opts`/`options` arguments received from the parent decorator in the `service.(function).call()` function. You can add to it but please remember to not change the received options. Changing the options could break the decorator above it.
```

Comment (markkaylor) [reply docs/docs/core/strapi/entity-service.md:24]: When you say you can add stuff to it, by "it" do you mean the options? or do you mean you can add more arguments to the call function?

File: docs/docs/core/strapi/entity-service.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+- when using `service.(function).call()` It is highly recommended for the first argument to be `this`. Not doing this will has a a result that the wrappers above won't be called since the call chain is broken
+- Wrappers are called on the lowest layer by the main decorator so no need to call them in your functions
+
+#### Main decorator
+When updating the main decorator, please make sure you follow this checklist:
+- When adding a new function please don't forget to add all wrapper functions to it. Not doing this could have as consequence that the wrapping of the input our output does not happen by decorators.
```
Comment (markkaylor) [inline docs/docs/core/strapi/entity-service.md:30]: > Not doing this could have as consequence that the wrapping of the input our output does not happen by decorators.

I don't really understand what's being said here.

File: docs/docs/core/strapi/entity-service.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+- You are recommend to add your changes to the output of `service.(function).call()` so you can return it with the output of the lower laying decorators.
+- It is best to use `opts`/`options` you get from the parent decorator into your service call into `service.(function).call()`. You can add stuff to it but please remember to not change the options you get. Since change them would have as an effect that it could break the decorator above it.
+- when using `service.(function).call()` It is highly recommended for the first argument to be `this`. Not doing this will has a a result that the wrappers above won't be called since the call chain is broken
+- Wrappers are called on the lowest layer by the main decorator so no need to call them in your functions
+
+#### Main decorator
+When updating the main decorator, please make sure you follow this checklist:
+- When adding a new function please don't forget to add all wrapper functions to it. Not doing this could have as consequence that the wrapping of the input our output does not happen by decorators. 
+- The convention is to use the function name as the wrapper action examples
```
Comment (markkaylor) [inline docs/docs/core/strapi/entity-service.md:31]:
```suggestion
- The convention is to use the function name as the wrapper action. For example:
```

**Review by markkaylor — COMMENTED** [review]
Thanks for this PR. I am not familiar with decorators in the entity service and these docs are definitely helpful. I left several small comments / suggestions for formatting and areas I found a bit hard to follow.

File: docs/docs/core/strapi/entity-service.md (hunk 7 above)
Comment (Boegie19) [reply docs/docs/core/strapi/entity-service.md:24]: Will also spesify this better by it I mean to the opts or options variables but you should not change anyting in it already since that would break layers that have run before it.

File: docs/docs/core/strapi/entity-service.md [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  - service
+  - entity-service
+---
+
+# Entity Service
+Strapi provides an Entity Service API, built on top of the Query Engine API (entity-manager in the Database Layer). The Entity Service is the layer that handles Strapi's complex data structures like components and dynamic zones, and uses the Query Engine API under the hood to execute database queries.
+## Decorators
+The Entity Service uses the decorator pattern so plugins and Strapi core packages can insert decorators to change the output of the Entity Service function.
+
+Examples of decorators are plugins/i18n and the core/upload.  You can decorate it like:
```
[the remark this answers is not in the document]
Comment (Boegie19) [reply docs/docs/core/strapi/entity-service.md:14]: plan for the code examples where to have examples of what are good and bad things to do and I will also link to the current code that is good.

**Comment by joshuaellis** [comment]
@Marc-Roig since you're most familiar with the entity service, i'm gonna hand this over to you 🙇🏼‍♀️
